Add/gutenboarding basic analytics

roo2 opened the pull request:
*Depends on https://github.com/Automattic/wp-calypso/pull/39392 being merged* I'll then need to rebase this onto `master`

This is a continuation of the calypso-analytics PR https://github.com/Automattic/wp-calypso/pull/39001
It adds examples of analytics to gutenboarding and:

* Pull the code that calculates the `last_pageview_path_with_count`  and `this_pageview_path_with_count` metadata into the calypso-analytics package.
* Add types to the calypso-analytics package
* Initialize analytics in gutenboarding
* Also added types for `User` and `UserData` ( ended up not using them in this PR but they are useful for editor autocompletion etc)
* Add example recordTracksPageView and recordTracksEvent calls to gutenboarding.

The next step will be to go through and add all of the analytics that we need in gutenboarding following the same pattern.

## Testing

- [x] Test existing page view events in calypso are setting `last_pageview_path_with_count` and `this_pageview_path_with_count` correctly
- [x] Test IE11 still attaches `?utm_` parameters (make sure there's no conflict with #39481)
- [x] Smoke test existing analytics in calypso
- [x] Test that `/gutenboarding` fires a pageView event and that `gutenboarding/signup` fires a calypso_gutenboarding_signup_start event with the pageview params set
Note: analytics in gutenboarding will not include *user* metadata such as number of sites created. This is because we will not have a *user* until an account is created

roo2 committed:
include meta data and initialize analytics for gutenboarding

sirreal commented:
I've added the "in progress" label, as I assumed this to be in progress 🙂
Using a draft PR will prevent code owners automatic review requests from being sent.

sirreal commented on packages/calypso-analytics/tsconfig.json:
```diff
@@ -12,7 +12,7 @@
 		"jsx": "react",
 		"declaration": true,
-		"declarationDir": "dist/types",
+		"declarationDir": "types",
```
Please use `dist/types` like the rest of the packages:

```suggestion
		"declarationDir": "dist/types",
```

sirreal commented on packages/calypso-analytics/package.json:
```diff
@@ -18,6 +18,7 @@
 		"types",
 		"src"
 	],
+	"types": "types",
```
```suggestion
	"types": "dist/types",
```

sirreal commented on packages/calypso-analytics/package.json:
```diff
@@ -0,0 +1,43 @@
+	},
+	"files": [
+		"dist",
+		"types",
```
```suggestion
```

sirreal commented on packages/calypso-analytics/tsconfig-cjs.json:
```diff
@@ -0,0 +1,10 @@
+		"declaration": false,
+		"declarationDir": null,
+		"outDir": "dist/cjs",
+		"tsBuildInfoFile": "../../.tsc-cache/data-stores-cjs"
```
```suggestion
		"tsBuildInfoFile": "../../.tsc-cache/calypso-analytics-cjs"
```

sirreal commented on packages/calypso-analytics/tsconfig.json:
```diff
@@ -0,0 +1,40 @@
+		"importHelpers": true,
+
+		"incremental": true,
+		"tsBuildInfoFile": "../../.tsc-cache/data-stores"
```
```suggestion
		"tsBuildInfoFile": "../../.tsc-cache/calypso-analytics"
```

sirreal commented on packages/load-script/package.json:
```diff
@@ -18,6 +18,7 @@
+	"types": "src/types",
 	"publishConfig": {
 		"access": "public"
 	},
```
This package needs `tslib` added to its `dependencies`.

sirreal commented on packages/calypso-analytics/global.d.ts:
```diff
@@ -0,0 +1,7 @@
+declare global {
+	interface Window {
+		_tkq: Array< Array< any > >;
+		BUILD_TIMESTAMP: string;
```
Is this intended to be a general purpose library? `BUILD_TIMESTAMP` global is very Calypso-specific, isn't it?

roo2 replied:
hmm yes, a more general approach would be to get the caller to pass in `BUILD_TIMESTAMP`... but that's not very nice for useability, I think it is better if we do just try to read the `window.BUILD_TIMESTAMP` if it's there, just to keep the function easy to use at the expense of leaking calypso concepts into the analytics library. What do you think?

sirreal commented on packages/calypso-analytics/global.d.ts:
```diff
@@ -0,0 +1,7 @@
+export {};
+declare global {
+	interface Window {
+		_tkq: Array< Array< any > >;
```
I believe it would be preferable to move this declaration to the module that adds `_tkq` to the window.

roo2 replied:
It seems like _tkq could be initialized in a few different places, each place where it is used, the code will initialize it if it's not defined with: `window._tkq = window._tkq || [];` It looks like calypso-analytics *is* the primary place it's initialized because the other places are all apps.

sirreal commented on packages/calypso-analytics/src/tracks.ts:
```diff
@@ -0,0 +1,253 @@
+
+if ( typeof window !== 'undefined' ) {
+	window._tkq = window._tkq || [];
+}
```
This would be a good place to add `_tkq` to the window type.

sirreal commented on packages/calypso-analytics/package.json:
```diff
@@ -0,0 +1,43 @@
+	"homepage": "https://github.com/Automattic/wp-calypso",
+	"license": "GPL-2.0-or-later",
+	"author": "Automattic Inc.",
+	"sideEffects": false,
```
This isn't accurate. At the very least `page-view-params` adds an event listener, `tracks` adds `_tkq` to the window object. I haven't audited the rest of the modules, but this should be fixed.

p-jackson replied:
@roo2 marked as resolved but this still doesn't look right to me. Unless I've misunderstood something, we can't mark this package as `sideEffects=false` because importing code from this package does indeed have side-effects (e.g. the event listener).
Does this need to be `sideEffects=true`?

roo2 replied:
Ahh sorry I accidentally hit mark as resolve instead of posting my comment here 🤐. 
It looks like `sideEffects` is used to determine if webpack is allowed to do tree shaking and not load files that aren't needed. If the module has `sideEffects` it means that there are implicit dependencies between functions. So "side effects" like attaching the event handler to window don't count as `sideEffects`, as long as other functions don't depend on it and it could safely be left out by the tree shaking build. 
https://stackoverflow.com/questions/49160752/what-does-webpack-4-expect-from-a-package-with-sideeffects-false

We used to have a cross function dependency on `window._tkq` being initialized at the top of the file, but I've refactored it so that `window._tkq` is not used directly, rather, the wrapping function `pushEventToTracksQueue` is used. I've also updated other places in calypso (`client/lib/analytics/index.js`) to use `pushEventToTracksQueue`. This way there are no implicit dependencies (despite still using the global) and the tree shaker can trace down to include only the functions that it needs.

So yes I believe it is correct to set `sideEffects:false` now, despite the code having "side effects".

p-jackson replied:
> "side effects" like attaching the event handler to window don't count as sideEffects, as long as other functions don't depend on it and it could safely be left out by the tree shaking build

Hmmm I hadn't thought of it like that 🤔

I guess in theory it shouldn't make a difference whether there's a latent window listener there or not. But I'm worried it's the sort of thing that'd be really hard to track down if it _did_ cause a difference between dev and prod builds. I'm thinking of cases where it was tricky fixing a Renovate PR because the modules getting updated had side effects.

It's not what I'd expect from something that's labelled side-effect free, but I see from the SO post that there's a fuzzy line since things like `console.log`s are generally ok. Do you have any opinions on where to draw the line @jon?

Another way would be to add the event listener in `initializeAnalytics()`, but that'd only work if it was being called early enough.

roo2 replied:
hmm It looks like setting `sideEffects=true` might be a safer and clearer option and I think that this module is so small and all in use that tree shaking wouldn't really make a difference, I'll just set it to true I think.

p-jackson replied:
Sweet, you did bring up a good point though. I'm interested in figuring out where we should be drawing that line in general.

sirreal replied:
You can also set it to a sideEffects to an array of files that have sideEffects, but just `true` is probably enough here. We aren't likely to make big gains here.

sirreal commented on packages/calypso-analytics/src/utils/current-user.ts:
```diff
@@ -0,0 +1,62 @@
+/**
+ * External dependencies
+ */
+import { UserData } from '../../../../client/lib/user/user';
```
Will this work? Do we want it to work? Generally, I wouldn't expect packages to import anything from the client app.

roo2 replied:
yea that was a bit crazy I took out that reference.

sirreal reviewed (commented):
Thanks for working on this. I've left a lot of comments and suggestions.

This seems to contain a lot of unrelated changes which makes review a bit more difficult. Was that intended?

roo2 commented:
Thanks for the reviews @sirreal! I've fixed most of the feedback and asked a followup question about the other. I also fixed the base so that the unrelated changes aren't included.

sirreal reviewed (commented):
Another round of drive-by review, I haven't been testing or looking closely at the code.

sirreal commented:
I've just noticed that this is "in progress" so perhaps my reviews were premature 😅 Hope they're helpful!

roo2 commented on packages/calypso-analytics/src/tracks.ts:
```diff
@@ -232,10 +251,15 @@ export function recordTracksPageView( urlPath, params ) {
 			urlParams &&
 			Array.from( urlParams.entries() ).filter( ( [ key ] ) => key.startsWith( 'utm_' ) );
-		const utmParams = utmParamEntries ? Object.fromEntries( utmParamEntries ) : {};
+		const utmParams = utmParamEntries ? ( Object as any ).fromEntries( utmParamEntries ) : {};
```
The fromEntries property isn't in the Typescript definitions for Object so typescript complains, it was only added recently https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/fromEntries. (I tested this does work in IE11 though, we must be transpiling it to something else)

roo2 replied:
this is fixed now, using your `ESNEXT` change in tsconfig

roo2 commented on client/landing/gutenboarding/onboarding-block/vertical-select/index.tsx:
```diff
@@ -138,6 +139,10 @@ const VerticalSelect: FunctionComponent< StepProps > = ( {
 
+	useEffect( () => {
+		recordTracksPageViewWithPageParams( '/gutenboarding' );
+	}, [] );
```
nice, I'll look into that

roo2 replied:
I tried to get this working but useLocation().pathname returns only the pathname within the current router. So here it returns '/about' when I want '/gutenboarding/about'. I looked for ways to get the '/gutenboarding' component of the path (the `BrowserRouter.basename` defined in `client/landing/gutenboarding/index.tsx:47`) but I couldn't find anything suitable beyond using `window.location.pathname` I think at this stage we should just hardcode it here. The main goal of the PR is to prove the new tracks API with examples. In the future we might do something slightly different, e.g. record every gutenboarding `calypso_page_view` from within `gutenboarding/index.js`.

roo2 replied:
update on this issue, I've moved page view tracking into the `Gutenboarding` component so each `gutenboarding/*` page view will be tracked automatically (thanks Philip)

roo2 commented on packages/load-script/package.json:
```diff
@@ -18,12 +18,14 @@
 		"@babel/runtime": "^7.4.4",
-		"debug": "^4.0.0"
+		"debug": "^4.0.0",
+		"tslib": "^1.10.0"
```
ahhh I see

roo2 commented on client/lib/user/user.d.ts:
```diff
@@ -0,0 +1,55 @@
+/**
```
Ok, you're right

> I've just noticed that this is "in progress" so perhaps my reviews were premature 😅 Hope they're helpful!

No, reviews are good :D sorry I forgot to update the label!

roo2 commented:
> I've just noticed that this is "in progress" so perhaps my reviews were premature 😅 Hope they're helpful!

No, reviews are good :D sorry I forgot to update the label!

roo2 commented on packages/calypso-analytics/src/tracks.ts:
```diff
@@ -11,8 +13,15 @@ import { loadScript } from '@automattic/load-script';
+	interface Window {
+		_tkq: Array< Array< any > >;
+	}
+}
```
Nice typescript foo 👍 
hmm this has exposed quite a few places where we weren't checking for `window` to be defined.. I wonder under what circumstances `window` would be undefined...? Some of this code is tested via `client/lib/analytics/test/index.js` and that test code also assumes that `window` exists. I can't think of any reason `window` would be undefined 😂.

p-jackson replied:
Some tests run in a `jsdom` environment and some don't, so just because all tests currently assume `window` is available, doesn't mean there won't be tests in the future that don't have `jsdom` but would still like to import this package. Adding `jsdom` to tests does slow them down a little.

roo2 replied:
ahh ok that makes sense, yea this code ideally shouldn't depend on jsdom for testing but it looks like it does for now. I'll just leave those guards around window.

roo2 commented on packages/calypso-analytics/src/page-view-params.ts:
```diff
@@ -0,0 +1,36 @@
+// we use this variable to track URL paths submitted to analytics.pageView.record
```
haha yep now the git blame will lead back to me 😂 

"but it's not the cleanest code - sorry" 
~ Eru Penkman (Approved by @sirreal)

I basically just tested in the browser that it produces the same behavior as it previously did and moved it all into it's own file.

p-jackson replied:
I see now, this comment was just copy-pasted from `lib/analytics`. Not sure what the nasty coupling is that it's refering to.

`analytics.pageView.record` and `analytics.pageLoading.record` are both in in `lib/analytics`, so maybe we should just leave this code there? Can we just reimplement the parts that are needed for `recordTracksPageViewWithPageParams` to work as expected?

roo2 replied:
Those comments seem to be incorrect/out of date. There is no `pageLoading.record` function, it isn't even in the git history, I checked out master and couldn't find any other references to the`mostRecentUrlPath` or `pathCounter` variables (appart from `analytics.timing.record`).

There is a coupling to `analytics.timing.record` because `analytics.timing.record` does re-use `mostRecentUrlPath` to submit its events. I think that's not ideal but it does it via the getMostRecentUrlPath function so at least it is explicit.

I updated the comments to reflect how the variables are being used.

p-jackson replied:
Thanks. The safest thing is probably to leave it as it is for now. If there's someone that "owns" `analytics.timing.record` then we could ask what the GA timings are being used for and if there's a cleaner way we could implement it. Maybe someone in @Automattic/team-calypso knows who's using the GA timing measurements? Seems like they're related to a config flag called `perfmon`.

Lets not do any of that in this PR though.

roo2 replied:
ahh do you mean to leave the PR as it is now or to leave this page_view params code in `lib/analytics/index.js`? 

It looks like @gravityrail wrote these comments about tracking `last_pageview_path_with_count` and `this_pageview_path_with_count` originally in 2016. I wonder if he remembers?

roo2 commented on client/landing/gutenboarding/index.tsx:
```diff
@@ -19,12 +20,24 @@ import accessibleFocus from 'lib/accessible-focus';
+		// User is left undefined here because the user account will not be created
+		// until after the user has completed the gutenboarding flow.
+		// This also saves us from having to pull in lib/user/user and it's dependencies.
+		initializeAnalytics( undefined, generateGetSuperProps() );
```
phew I think that is a different `User` :joy: I was including the `analytics` user returned by `client/lib/user/user.js`, but I wonder if the `client/lib/user/user.js` would also work 🤔. 

Yea I'm with you on getting more type info in and hopefully this `@automattic/calypso-types` is the right place for those unused `client/lib/user/user.js` type definitions I wrote.

I wrote a note to come back to this :)

roo2 commented on packages/calypso-analytics/package.json:
```diff
@@ -5,19 +5,19 @@
-	"sideEffects": false,
-	"main": "dist/cjs/index.js",
-	"module": "dist/esm/index.js",
+	"sideEffects": true,
```
hmm ok so `sideEffects` is a flag controlling whether webpack can safely prune unused exported functions or not. If it's `true` than all functions will always be included in the bundle because webpack can't work out if they are necessary just for their "sideEffects". https://stackoverflow.com/questions/49160752/what-does-webpack-4-expect-from-a-package-with-sideeffects-false

Our only side effect is initializing `window._tkq` when the library is loaded; `client/lib/analytics/index.js` does in fact depend on this having happened.

I've actually gone and contained all usage of `window._tkq` inside a new `pushEventToTracksQueue()` function, and updated the dependent code in `client/lib/analytics/` to use this. 

So now the code still has the "Side Effect" of adding to the `window` object, but it doesn't have `sideEffects` because there are no implicit dependencies between functions :blush:.

There are other places in `apps/` where we use `window._tkq` but fortunately these places check to ensure that it's been initialized. I probably should update them to all use `pushEventToTracksQueue` but I don't want to introduce the dependency on `@automattic/calypso-analytics` to those apps right now 😋

sirreal commented on packages/calypso-analytics/tsconfig.json:
```diff
@@ -0,0 +1,35 @@
+{
+	"compilerOptions": {
+		"target": "ES5",
+		"lib": [ "es6", "DOM", "DOM.Iterable" ],
```
Regarding https://github.com/Automattic/wp-calypso/pull/39451/files#r381582967 - let's try this:

```suggestion
		"lib": [ "DOM", "ESNEXT" ],
```

[`Object.fromEntries` is available](https://unpkg.com/browse/typescript@3.7.5/lib/lib.es2019.object.d.ts), it's a question of finding the right configuration.

roo2 committed:
fix typescript config

ramonjd commented:
Sorry, this PR is too big for me to give a meaningful review today. Thanks to @sirreal for picking through the code, and for the informative reviews!

But I did go through Gutenboarding - 

`calypso_page_view` fires for `gutenboarding/about` and `calypso_gutenboarding_signup_start` fires correctly 👍 

Also smoke testing analytics throughout Calypso to look for regressions. All page load and click tracking events are firing, and I can't see any obvious issues.

p-jackson commented on client/landing/gutenboarding/onboarding-block/vertical-select/index.tsx:
```diff
@@ -139,6 +140,10 @@ const VerticalSelect: FunctionComponent< StepProps > = ( {
 
+	useEffect( () => {
+		recordTracksPageViewWithPageParams( '/gutenboarding/about' );
+	}, [] );
```
Just a quick comment for now: I wonder if this could be done in a general way rather than having to call `recordTracksPageViewWithPageParams` on mount of each component. Plus it's susceptible to breaking when we make changes to the route.

I was thinking about listening to route updates instead of component mounts e.g. adding something like this to the `<Gutenboard>` component:

```typescript
import { useParams, useHistory } from 'react-router-dom';

const { location: { pathname } } = useHistory();
useEffect( () => {
	recordTracksPageViewWithPageParams( `/gutenboarding${pathname}` );
}, [ pathname ] );
```

Adding it to `gutenboard.tsx` instead of `index.tsx` means it will only be triggered by routes that we actually support i.e. no `/gutenboarding/abcxyz` track events.

Another thing it will do differently is send the locale slug as part of the event too. I'm not sure whether that's how `recordTracksPageViewWithPageParams` is supposed to be used, but I assume so.

What do you think?

p-jackson replied:
Just checked the old signup framework, and yes it's supposed to include the language slug.

roo2 replied:
nice, that looks good, I'll add that code and test it, thanks.

roo2 committed:
use recordPageView on every gutenboarding page

p-jackson commented on packages/calypso-analytics/src/tracks.ts:
```diff
@@ -27,21 +37,20 @@ const EVENT_NAME_EXCEPTIONS = [
+function createRandomId( randomBytesLength = 9 ): string {
+	if ( typeof window === 'undefined' ) {
+		return '';
+	}
```
If `window` isn't defined maybe this should go down the same code path as when `window.crypto` isn't defined.

roo2 replied:
ahh I can't do that because in the case where `window.crypto` is undefined, it still uses the `window.btoa` function on line `:62`

p-jackson replied:
lol so it does, missed that 😂

p-jackson commented on packages/calypso-analytics/src/utils/hash-pii.ts:
```diff
@@ -0,0 +1,15 @@
+ * Hashes users' Personally Identifiable Information using SHA256
+ *
+ * @param {string|number} data Data to be hashed
+ * @returns {string} SHA256 in hex string format
```
Now that this is TS we can delete the `{...}` in the jsdoc

p-jackson commented on packages/calypso-analytics/src/utils/current-user.ts:
```diff
@@ -0,0 +1,60 @@
+ * Sets current user, (stored in javascript memory).
+ *
+ * @param {object} currentUser the user data for the current user
+ * @returns {object|undefined} Current user.
```
We don't need the {...} in the jsdoc here

p-jackson commented on packages/calypso-analytics/src/utils/current-user.ts:
```diff
@@ -0,0 +1,60 @@
+	ID: number;
+	username: string;
+	email: string;
+	hashedPii?: CurrentUserHashedPii;
```
Are there any cases where `hashedPii` can be missing? If we make it non-optional it would catch any cases where we forget to set.

p-jackson commented on packages/calypso-analytics/src/utils/current-user.ts:
```diff
@@ -0,0 +1,60 @@
+ * @returns {object|undefined} Current user.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function setCurrentUser( currentUser: any ): CurrentUser | undefined {
```
I was wondering if it'd be better to just define a wee interface for this function's parameter type. Something like
```typescript
interface SetCurrentUserParam {
	ID: string;
	username: string;
	email: string;
}
```

It doesn't need to be exported.

p-jackson commented on packages/calypso-analytics/src/utils/current-user.ts:
```diff
@@ -0,0 +1,60 @@
+export function setCurrentUser( currentUser: any ): CurrentUser | undefined {
+	if (
+		! currentUser.ID ||
+		isNaN( Number( currentUser.ID ) ) ||
```
This logic changed very subtly from what it used to be in JS.

```suggestion
		isNaN( parseInt( currentUser.ID, 10 ) ) ||
```

Without knowing much about the format of the `ID` string, I'd say it's safer to stick with `parseInt` so we can use the explicit radix param.

roo2 replied:
Good catch!

p-jackson commented on client/landing/gutenboarding/gutenboard.tsx:
```diff
@@ -56,6 +57,14 @@ export function Gutenboard() {
+	} = useHistory();
+
+	useEffect( () => {
+		recordTracksPageViewWithPageParams( `/gutenboarding${ pathname }` );
```
Just noticed that when I navigate to `/gutenboarding` and look at the `calypso_page_view` event in the network inspector, it's sending a trailing slash in the path i.e. `path=/gutenboarding/`. Not sure how much it matters (I don't know every way tracks data is used 😅), it might be annoying for folks designing their funnels using tracks data. But fwiw events from the `/start` onboarding flow don't have a trailing slash.

roo2 replied:
hmm it's different but I don't think we should add code to change it unless it does turn out to cause any issues for the analytics people, the other routes like `/gutenboarding/start` don't have a trailing `/`, so the `pathname` must be `/` here to give us `/gutenboarding/`. I think it's alright as long as they can track the unique URLs

p-jackson reviewed (commented):
Phew that's a lot of analytics 😅 Breaking the PR up into:

1. changes related to `*_with_count` params
2. converting to TypeScript
3. record events in gutenboarding

would make the reviewing a bit easier. Particularly because there might be a lot more discussion around (1) and how we actually want analytics to work in the new package world (which is a good thing to talk about), but that discussion doesn't need to hold up (3).

roo2 committed:
Fix some issues from PR feedback

roo2 commented:
> Phew that's a lot of analytics 😅 Breaking the PR up into:
> 
>     1. changes related to `*_with_count` params
> 
>     2. converting to TypeScript
> 
>     3. record events in gutenboarding
> 
> 
> would make the reviewing a bit easier. Particularly because there might be a lot more discussion around (1) and how we actually want analytics to work in the new package world (which is a good thing to talk about), but that discussion doesn't need to hold up (3).

Yes, I should have done the typescript conversion seperately to make this easier to review, I think that *3)*, recording the `calypso_page_view` event in gutenboarding is dependent on *1)*, pulling the page view params into `calypso-analytics`, I haven't made changes to that functionality works (apart from that integer conversion which you found, oops) but just pulled it all into one place to be reused. So yea I don't know if we want to change the way the `*_with_count` stuff works, now would be a good opportunity to change it but for now it's just been moved.

So it would have been good but I'd rather not pull the PR into separate parts now that it's mostly reviewed 😅

roo2 commented:
hmm yea, I signed up for a new site with the free domain `erutestredirectsite.design.blog` and then set it to my default site. when I set the route to `/`, I got this page:

<img width="706" alt="Screen Shot 2020-02-25 at 4 13 18 PM" src="https://user-images.githubusercontent.com/22446385/75217079-1b49b300-57ea-11ea-940a-6e12890a3486.png">

roo2 committed:
update comments

p-jackson commented:
Been manually testing the changes in Calypso and the `calypso_page_view` and GA timing events are working as expected (i.e. the same way they work on `master`).

Testing GA was a little tricky, but I did it by adding another `debug` line to `analytics.timing.record()` and enabling `perfmon` in `development.json`.

I think the last remaining thing in this PR is to do with `sideEffects` in the `package.json`. I think it's changed between `false` and `true` a couple times during the history of the PR so I'm struggling to figure out the conversation that's already happened 😅
@roo2 is the plan to switch it to `sideEffects=true`

roo2 commented:
Thanks @p-jackson for the reviews and especially for manually testing `analytics.timing.record`!

I think that we should have `sideEffects=false` because that flag is used to track whether tree shaking can be used on the library, even if there are side effects like the one you pointed out  https://github.com/Automattic/wp-calypso/pull/39451#discussion_r384204666

I really struggled to find the conversation too and ended up writing another comment, I installed this user script via tampermonkey to add an "Expand, Expand" button to github to show all conversations 😂 https://github.com/findepi/user.js/blob/master/github-expand-expand-expand.user.js

roo2 committed:
set sideEffects=true after PR discusion

p-jackson reviewed (approved):
This LGTM

The failing tests look like they have something to do with Customer Home/Stats changes, which I thought would have been fixed by #39653. Either way, I've certainly been able to navigate to the stats page in my testing, so the failures look unrelated.

p-jackson commented:
> "Expand, Expand"

Nice! Added.
